shopfloor/actions/search.py

```diff
@@ -157,3 +158,18 @@ def generic_packaging_from_scan(self, barcode):
         return model.search(
             [("barcode", "=", barcode), ("product_id", "=", False)], limit=1
         )
+
+    def delivered_picking_from_scan(self, barcode, use_origin=False):
+        picking = self.packaging_from_scan(barcode)
+        if picking:
+            return picking
+        model = self.env["stock.picking"]
+        if use_origin:
+            source_document_domain = [
+                # We could have the same origin for multiple transfers
+                # but we're interested only in the "done" ones.
+                ("origin", "=", barcode),
+                ("state", "=", "done"),
+            ]
+            return model.search(source_document_domain)
```

Comment:
Some remarks regarding the origin:
* we need to be able to pre-process the barcode. In our case, the origin is prefixed by the transporter and this prefix must be dropped to match a picking
* I would have added a domain to the method and include it in the document origin search. For picking, the domain would be type=incoming + state=assigned. For return, the domain would be type=outgoing + state=done. Then you would not need this new handler.
* a picking can have multiple origins separated by a comma. You could have a reception that concerns multiple POs. You can have a delivery to return that concerns multiple SOs. In those cases, the search on "=" won't work. I wonder if the origin must not be on moves instead of picking. This is because after matching the picking, you still have to match the move based on the scanned product/packaging/lot but you cannot mix the origins. You then would have an handler for the picking only on its name and another handler on the move origin